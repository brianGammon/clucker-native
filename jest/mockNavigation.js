/* @flow */
type ReturnVals = {
  [methodName: string]: {
    [param: string]: string,
  },
};

type MockNavigation = {
  getParam: (param: string, defaultId: string) => string,
};

const mockNavigation = (returnVals: ReturnVals) => ({
  getParam: jest.fn((param: string, defaultId: string) => {
    if (returnVals.getParam && returnVals.getParam[param]) {
      return returnVals.getParam[param];
    }
    return defaultId;
  }),
}: MockNavigation);

export default mockNavigation;
