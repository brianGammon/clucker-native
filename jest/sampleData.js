const sampleData1 = {
  userSettings: {
    key: 'user1',
    data: {
      currentFlockId: 'flock1',
    },
  },
  flocks: {
    data: {
      flock1: {
        name: 'Test Flock 1',
        ownedBy: 'user1',
      },
      flock2: {
        name: 'Test Flock 2',
        ownedBy: 'user2',
      },
    },
  },
  chickens: {
    data: {
      chicken1: {
        name: 'The First Chicken',
      },
      chicken2: {
        name: 'The Second Chicken',
      },
    },
  },
  eggs: {
    data: {
      egg1: {
        chickenId: 'chicken1',
        chickenName: 'Bossie',
        damaged: false,
        date: '2018-08-01',
        modified: '2018-01-12T18:05:54.806Z',
        notes:
          'This was the record setting egg from May 2017, real egg was 66.1g on this day',
        userId: 'MVQUEcoTaVUhUdo2ZS0zsIE6tn93',
        weight: 88.1,
      },
      egg2: {
        chickenId: 'chicken2',
        chickenName: 'Baby',
        damaged: false,
        date: '2017-09-24',
        modified: '2017-10-12T18:05:54.806Z',
        userId: 'MVQUEcoTaVUhUdo2ZS0zsIE6tn93',
        weight: 88.1,
      },
      egg3: {
        chickenId: 'chicken1',
        chickenName: 'Bossie',
        damaged: false,
        date: '2018-08-02',
        modified: '2018-08-02T18:05:54.806Z',
        notes: '',
        userId: 'MVQUEcoTaVUhUdo2ZS0zsIE6tn93',
        weight: 68.1,
      },
    },
  },
};

export default sampleData1;
