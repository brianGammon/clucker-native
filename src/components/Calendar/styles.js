import { StyleSheet } from 'react-native';

const label = {
  color: 'grey',
  fontWeight: 'bold',
  fontSize: 17,
};

const cell = {
  width: '14.28571%',
  alignItems: 'center',
};

const rowContainer = {
  flexDirection: 'row',
};

export default StyleSheet.create({
  section: {
    marginTop: 10,
  },
  labelContainer: {
    ...rowContainer,
  },
  labelCell: {
    ...cell,
    borderBottomWidth: 1,
  },
  outside: {
    backgroundColor: '#ddd',
  },
  label,
  calendarContainer: {
    ...rowContainer,
    flexWrap: 'wrap',
  },
  dayCell: {
    ...cell,
    borderRightWidth: 1,
    borderBottomWidth: 1,
    minHeight: 60,
  },
  innerDayCell: {
    padding: 2,
    flex: 1,
    width: '100%',
  },
  sunday: {
    borderLeftWidth: 1,
  },
  dayLabel: {
    ...label,
    fontWeight: 'normal',
    alignSelf: 'flex-start',
  },
  eggCount: {
    alignSelf: 'center',
  },
  statsLabel: {
    fontWeight: 'bold',
    fontSize: 17,
  },
});
