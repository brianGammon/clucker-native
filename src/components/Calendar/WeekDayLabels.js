import React from 'react';
import { View, Text } from 'native-base';
import styles from './styles';

const labels = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
const WeekDayLabels = () => (
  <View style={styles.labelContainer}>
    {labels.map(day => (
      <View key={day} style={styles.labelCell}>
        <Text style={styles.label}>{day}</Text>
      </View>
    ))}
  </View>
);

export default WeekDayLabels;
