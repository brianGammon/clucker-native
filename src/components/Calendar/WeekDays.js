import React from 'react';
import moment from 'moment';
import { TouchableOpacity } from 'react-native';
import { View, Text, Badge } from 'native-base';
import { type CalendarData, type Navigation } from '../../types';
import styles from './styles';

type Props = {
  navigation: Navigation,
  calendarData: CalendarData,
  month: string,
};

const isSunday = day => moment(day).day() === 0;
const isInRange = (day, month) => moment(day).month() === moment(month).month();

const WeekDays = ({ calendarData, month, navigation }: Props) => (
  <View style={styles.calendarContainer}>
    {Object.keys(calendarData || {}).map(day => (
      <TouchableOpacity
        key={day}
        onPress={() => navigation.navigate('Day', { date: day })}
        style={[
          styles.dayCell,
          isSunday(day) && styles.sunday,
          !isInRange(day, month) && styles.outside,
        ]}
      >
        <View style={styles.innerDayCell}>
          <Text style={styles.dayLabel}>{moment(day).format('D')}</Text>
          {calendarData[day].total && (
            <Badge success style={styles.eggCount}>
              <Text>{calendarData[day].total}</Text>
            </Badge>
          )}
        </View>
      </TouchableOpacity>
    ))}
  </View>
);

export default WeekDays;
