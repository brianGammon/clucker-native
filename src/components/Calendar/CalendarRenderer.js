/* @flow */
import * as React from 'react';
import {
  View, Container, Content, Text,
} from 'native-base';
import { type Chicken, type CalendarData, type FlockStats } from '../../types';
import Header from '../Header';
import styles from './styles';
import WeekDayLabels from './WeekDayLabels';
import DateSwitcher from '../DateSwitcher';
import Line from '../Line';
import WeekDays from './WeekDays';
import Leaderboard from '../Leaderboard';
import HeaviestEgg from '../HeaviestEgg';
import HelpMessage from '../HelpMessage';

type Props = {
  navigation: any,
  stats: FlockStats,
  calendarData: CalendarData,
  chickens: {
    [chickenId: string]: Chicken,
  },
  dates: {
    now: string,
    date: string,
    previousDate: string,
    nextDate: string,
  },
};

const CalendarRenderer = ({
  navigation,
  calendarData,
  dates,
  chickens,
  stats,
}: Props) => (
  <Container>
    <Header title="Calendar" eggButton />
    <Content padder>
      <DateSwitcher
        mode="month"
        navigation={navigation}
        dates={dates}
        eggCount={stats && stats.total}
      />

      <View style={styles.section}>
        <WeekDayLabels />
        <WeekDays
          calendarData={calendarData}
          month={dates.date}
          navigation={navigation}
        />
      </View>

      {stats && (
        <View>
          <Line />
          <Leaderboard stats={stats} chickens={chickens} mode="month" />
          <Line />
          {stats.heaviest && (
            <HeaviestEgg
              heaviest={stats.heaviest}
              chickenName={
                (chickens[stats.heaviest.chickenId]
                  && chickens[stats.heaviest.chickenId].name)
                || 'Unnamed Hen'
              }
            />
          )}
        </View>
      )}
      {!stats && (
        <View>
          <Line />
          <HelpMessage>
            <Text>No eggs logged yet for this period</Text>
          </HelpMessage>
        </View>
      )}
    </Content>
  </Container>
);

export default CalendarRenderer;
