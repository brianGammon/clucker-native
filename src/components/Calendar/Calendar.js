import * as React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import CalendarRenderer from './CalendarRenderer';
import flockStatsSelector from '../../selectors/flockStatsSelector';
import { nowAsMoment, dateStringAsMoment } from '../../utils/dateHelper';
import { type Chicken } from '../../types';

type Props = {
  navigation: any,
  stats: any,
  chickens: {
    [chickenId: string]: Chicken,
  },
  dates: {
    date: string,
    previousDate: string,
    nextDate: string,
  },
};

class Calendar extends React.Component<Props> {
  render() {
    const {
      navigation, stats, dates, chickens,
    } = this.props;
    return (
      <CalendarRenderer
        navigation={navigation}
        stats={stats}
        chickens={chickens}
        dates={dates}
      />
    );
  }
}

const mapStateToProps = ({ eggs, chickens }, { navigation }) => {
  const date = navigation.getParam('date', nowAsMoment().format('YYYY-MM'));
  const currDate = moment.utc(date);
  const previousDate = currDate
    .subtract(1, 'months')
    .format('YYYY-MM')
    .toString();
  let nextDate = currDate
    .add(2, 'months')
    .format('YYYY-MM')
    .toString();
  if (dateStringAsMoment(nextDate).isAfter(nowAsMoment())) {
    nextDate = null;
  }
  return {
    stats: flockStatsSelector(eggs.data, date),
    chickens: chickens.data,
    dates: {
      date,
      previousDate,
      nextDate,
    },
  };
};

export default connect(mapStateToProps)(Calendar);
