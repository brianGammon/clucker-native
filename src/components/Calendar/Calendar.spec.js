import React from 'react';
import configureStore from 'redux-mock-store';
import sampleData from '../../../jest/sampleData';
import mockNavigation from '../../../jest/mockNavigation';
import Calendar from '.';

describe('Calender component:', () => {
  const middlewares = [];
  const mockStore = configureStore(middlewares);
  const store = mockStore(sampleData);

  test('Should render', () => {
    const paramReturned = {
      getParam: {
        date: '2018-08',
      },
    };
    const navigation = mockNavigation(paramReturned);
    const wrapper = shallow(
      <Calendar store={store} navigation={navigation} />,
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });
});
