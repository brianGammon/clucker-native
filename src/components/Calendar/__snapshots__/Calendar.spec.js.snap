// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calender component: Should render 1`] = `
<CalendarRenderer
  calendarData={
    Object {
      "2018-04-29": Object {},
      "2018-04-30": Object {},
      "2018-05-01": Object {
        "byChicken": Object {
          "chicken1": 1,
          "chicken2": 1,
          "chicken4": 1,
        },
        "total": 3,
      },
      "2018-05-02": Object {
        "byChicken": Object {
          "chicken1": 1,
          "chicken2": 1,
          "chicken4": 1,
        },
        "total": 3,
      },
      "2018-05-03": Object {
        "byChicken": Object {
          "chicken1": 1,
          "chicken2": 1,
        },
        "total": 2,
      },
      "2018-05-04": Object {
        "byChicken": Object {
          "chicken1": 1,
          "chicken4": 1,
        },
        "total": 2,
      },
      "2018-05-05": Object {
        "byChicken": Object {
          "chicken1": 1,
          "chicken2": 1,
        },
        "hasNote": true,
        "total": 2,
      },
      "2018-05-06": Object {
        "byChicken": Object {
          "chicken1": 1,
          "chicken4": 1,
        },
        "hasNote": true,
        "total": 2,
      },
      "2018-05-07": Object {
        "byChicken": Object {
          "chicken1": 1,
          "chicken4": 1,
        },
        "total": 2,
      },
      "2018-05-08": Object {
        "byChicken": Object {
          "chicken1": 1,
          "chicken2": 1,
        },
        "total": 2,
      },
      "2018-05-09": Object {
        "byChicken": Object {
          "chicken1": 1,
          "chicken2": 1,
          "chicken4": 1,
        },
        "total": 3,
      },
      "2018-05-10": Object {
        "byChicken": Object {
          "chicken2": 1,
          "chicken4": 1,
        },
        "hasNote": true,
        "total": 2,
      },
      "2018-05-11": Object {
        "byChicken": Object {
          "chicken1": 1,
          "chicken2": 1,
        },
        "total": 2,
      },
      "2018-05-12": Object {
        "byChicken": Object {
          "chicken1": 1,
          "chicken4": 1,
        },
        "hasNote": true,
        "total": 2,
      },
      "2018-05-13": Object {
        "byChicken": Object {
          "chicken1": 1,
        },
        "total": 1,
      },
      "2018-05-14": Object {},
      "2018-05-15": Object {},
      "2018-05-16": Object {},
      "2018-05-17": Object {},
      "2018-05-18": Object {},
      "2018-05-19": Object {},
      "2018-05-20": Object {},
      "2018-05-21": Object {},
      "2018-05-22": Object {},
      "2018-05-23": Object {},
      "2018-05-24": Object {},
      "2018-05-25": Object {},
      "2018-05-26": Object {},
      "2018-05-27": Object {},
      "2018-05-28": Object {},
      "2018-05-29": Object {},
      "2018-05-30": Object {},
      "2018-05-31": Object {},
      "2018-06-01": Object {},
      "2018-06-02": Object {},
    }
  }
  chickens={
    Object {
      "chicken1": Object {
        "breed": "Cuckoo Maran",
        "hatched": "2016-10-05",
        "name": "Chicken1",
        "photoPath": "uploads/user:user1/flock:flock1/1BD434AA-4A5D-40A9-A97A-C3D64FC8B841.jpeg",
        "photoUrl": "https://firebasestorage.googleapis.com/v0/b/cluckr-prod.appspot.com/o/uploads%2Fuser%3AvEFMHqRcozMSriuqeJNKlCXnGMr2%2Fflock%3Aflock1%2F1BD434AA-4A5D-40A9-A97A-C3D64FC8B841.jpeg?alt=media&token=029f7593-9ed3-481e-8dc6-5d3c38844180",
        "thumbnailPath": "uploads/user:user1/flock:flock1/thumb_1BD434AA-4A5D-40A9-A97A-C3D64FC8B841.jpeg",
        "thumbnailUrl": "https://firebasestorage.googleapis.com/v0/b/cluckr-prod.appspot.com/o/uploads%2Fuser%3AvEFMHqRcozMSriuqeJNKlCXnGMr2%2Fflock%3Aflock1%2Fthumb_1BD434AA-4A5D-40A9-A97A-C3D64FC8B841.jpeg?alt=media&token=689dfae8-c034-46ae-9984-db5c4ee3fe72",
      },
      "chicken2": Object {
        "breed": "Black Australorpe",
        "hatched": "2016-10-05",
        "name": "Chicken2",
        "photoPath": "uploads/user:user1/flock:flock1/8BF48E4E-F5EF-4DCC-91C1-746CF33586FB.jpeg",
        "photoUrl": "https://firebasestorage.googleapis.com/v0/b/cluckr-prod.appspot.com/o/uploads%2Fuser%3AvEFMHqRcozMSriuqeJNKlCXnGMr2%2Fflock%3Aflock1%2F8BF48E4E-F5EF-4DCC-91C1-746CF33586FB.jpeg?alt=media&token=0ae6a73d-7027-4d18-8ab2-3f40340496c5",
        "thumbnailPath": "uploads/user:user1/flock:flock1/thumb_8BF48E4E-F5EF-4DCC-91C1-746CF33586FB.jpeg",
        "thumbnailUrl": "https://firebasestorage.googleapis.com/v0/b/cluckr-prod.appspot.com/o/uploads%2Fuser%3AvEFMHqRcozMSriuqeJNKlCXnGMr2%2Fflock%3Aflock1%2Fthumb_8BF48E4E-F5EF-4DCC-91C1-746CF33586FB.jpeg?alt=media&token=0e0d4dea-5104-4cfe-9a49-f87d4b412e24",
      },
      "chicken3": Object {
        "breed": "Welsummer",
        "hatched": "2016-10-05",
        "name": "Chicken3",
        "photoPath": "uploads/user:user1/flock:flock1/1A2F5FA5-FE29-4D43-9E4A-E44F19CE44B9.jpeg",
        "photoUrl": "https://firebasestorage.googleapis.com/v0/b/cluckr-prod.appspot.com/o/uploads%2Fuser%3AvEFMHqRcozMSriuqeJNKlCXnGMr2%2Fflock%3Aflock1%2F1A2F5FA5-FE29-4D43-9E4A-E44F19CE44B9.jpeg?alt=media&token=54a9eb90-7a2a-4370-a030-7cb3fa3bb738",
        "thumbnailPath": "uploads/user:user1/flock:flock1/thumb_1A2F5FA5-FE29-4D43-9E4A-E44F19CE44B9.jpeg",
        "thumbnailUrl": "https://firebasestorage.googleapis.com/v0/b/cluckr-prod.appspot.com/o/uploads%2Fuser%3AvEFMHqRcozMSriuqeJNKlCXnGMr2%2Fflock%3Aflock1%2Fthumb_1A2F5FA5-FE29-4D43-9E4A-E44F19CE44B9.jpeg?alt=media&token=708ffb11-4a0a-4dab-9633-27a2c0c74d18",
      },
      "chicken4": Object {
        "breed": "Welsummer",
        "hatched": "2016-10-05",
        "name": "Chicken4",
        "photoPath": "uploads/user:user1/flock:flock1/9B6AF4A8-B948-4BF1-AEB6-9D78AD9A382E.jpeg",
        "photoUrl": "https://firebasestorage.googleapis.com/v0/b/cluckr-prod.appspot.com/o/uploads%2Fuser%3AvEFMHqRcozMSriuqeJNKlCXnGMr2%2Fflock%3Aflock1%2F9B6AF4A8-B948-4BF1-AEB6-9D78AD9A382E.jpeg?alt=media&token=77141d6f-bb85-4174-817b-ad2cad174315",
        "thumbnailPath": "uploads/user:user1/flock:flock1/thumb_9B6AF4A8-B948-4BF1-AEB6-9D78AD9A382E.jpeg",
        "thumbnailUrl": "https://firebasestorage.googleapis.com/v0/b/cluckr-prod.appspot.com/o/uploads%2Fuser%3AvEFMHqRcozMSriuqeJNKlCXnGMr2%2Fflock%3Aflock1%2Fthumb_9B6AF4A8-B948-4BF1-AEB6-9D78AD9A382E.jpeg?alt=media&token=baa3c541-626a-45e7-b75c-1eab00454476",
      },
    }
  }
  dates={
    Object {
      "date": "2018-05",
      "nextDate": "2018-06",
      "now": "2018-05",
      "previousDate": "2018-04",
    }
  }
  navigation={
    Object {
      "getParam": [MockFunction] {
        "calls": Array [
          Array [
            "date",
            "2018-05",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": "2018-05",
          },
        ],
      },
    }
  }
  stats={
    Object {
      "averageNumber": 1.8666666666666667,
      "averageWeight": 65.21428571428572,
      "eggsPerChicken": Object {
        "chicken1": 12,
        "chicken2": 8,
        "chicken4": 8,
      },
      "eggsPerPeriod": Object {
        "2018-05-01": Object {
          "byChicken": Object {
            "chicken1": 1,
            "chicken2": 1,
            "chicken4": 1,
          },
          "total": 3,
        },
        "2018-05-02": Object {
          "byChicken": Object {
            "chicken1": 1,
            "chicken2": 1,
            "chicken4": 1,
          },
          "total": 3,
        },
        "2018-05-03": Object {
          "byChicken": Object {
            "chicken1": 1,
            "chicken2": 1,
          },
          "total": 2,
        },
        "2018-05-04": Object {
          "byChicken": Object {
            "chicken1": 1,
            "chicken4": 1,
          },
          "total": 2,
        },
        "2018-05-05": Object {
          "byChicken": Object {
            "chicken1": 1,
            "chicken2": 1,
          },
          "hasNote": true,
          "total": 2,
        },
        "2018-05-06": Object {
          "byChicken": Object {
            "chicken1": 1,
            "chicken4": 1,
          },
          "hasNote": true,
          "total": 2,
        },
        "2018-05-07": Object {
          "byChicken": Object {
            "chicken1": 1,
            "chicken4": 1,
          },
          "total": 2,
        },
        "2018-05-08": Object {
          "byChicken": Object {
            "chicken1": 1,
            "chicken2": 1,
          },
          "total": 2,
        },
        "2018-05-09": Object {
          "byChicken": Object {
            "chicken1": 1,
            "chicken2": 1,
            "chicken4": 1,
          },
          "total": 3,
        },
        "2018-05-10": Object {
          "byChicken": Object {
            "chicken2": 1,
            "chicken4": 1,
          },
          "hasNote": true,
          "total": 2,
        },
        "2018-05-11": Object {
          "byChicken": Object {
            "chicken1": 1,
            "chicken2": 1,
          },
          "total": 2,
        },
        "2018-05-12": Object {
          "byChicken": Object {
            "chicken1": 1,
            "chicken4": 1,
          },
          "hasNote": true,
          "total": 2,
        },
        "2018-05-13": Object {
          "byChicken": Object {
            "chicken1": 1,
          },
          "total": 1,
        },
      },
      "firstEgg": "2018-05-01",
      "heaviest": Object {
        "chickenId": "chicken1",
        "chickenName": "Chicken1",
        "date": "2018-05-01",
        "modified": "2018-05-01T19:09:21.715Z",
        "userId": "user1",
        "weight": "71.4",
      },
      "mostEggs": "chicken1",
      "total": 28,
    }
  }
/>
`;
